src/safe_mpc/cost_definition.py

```diff
@@ -0,0 +1,125 @@
+import scipy.linalg as lin
+
+
+# class to manage cost settings: until now only Q*error_ee*Q + R*u*R
+class AbstractCostSettings():
+    def __init__(self, controller):
+        self.controller = controller
+        self.cost_type = None
+
+    def set_ocp_cost_type(self):
+        self.controller.ocp.cost.cost_type_0 = self.cost_type
+        self.controller.ocp.cost.cost_type = self.cost_type
+        self.controller.ocp.cost.cost_type_e = self.cost_type
+
+class ReachTargetNLS(AbstractCostSettings):
+    def __init__(self,controller):
+        super().__init__(controller)
+        self.cost_type = 'NONLINEAR_LS'
+        self.set_ocp_cost_type()
+
+        self.controller.ocp.model.cost_y_expr_0 = cs.vertcat(self.controller.model.t_glob - self.controller.ee_params,self.controller.model.u)
+        self.controller.ocp.cost.W_0 = lin.block_diag(self.controller.model.params.Q_weight*np.eye(self.controller.model.t_glob.shape[0]),self.controller.model.params.R_weight*np.eye(self.controller.model.nu))
+        self.controller.ocp.cost.yref_0 = np.hstack((np.zeros((self.controller.model.t_glob.shape[0],)), np.zeros(self.controller.model.nu)))
+
+        self.controller.ocp.model.cost_y_expr = cs.vertcat(self.controller.model.t_glob - self.controller.ee_params,self.controller.model.u)
+        self.controller.ocp.cost.W = lin.block_diag(self.controller.model.params.Q_weight*np.eye(self.controller.model.t_glob.shape[0]),self.controller.model.params.R_weight*np.eye(self.controller.model.nu))
+        self.controller.ocp.cost.yref = np.hstack((np.zeros((self.controller.model.t_glob.shape[0],)), np.zeros(self.controller.model.nu)))
+
+        self.controller.ocp.model.cost_y_expr_e = cs.vertcat(self.controller.model.t_glob - self.controller.ee_params)
+        self.controller.ocp.cost.W_e = lin.block_diag(self.controller.model.params.Q_weight*np.eye(self.controller.model.t_glob.shape[0]))
+        self.controller.ocp.cost.yref_e = np.zeros((self.controller.model.t_glob.shape[0],))
+
+        self.traj = np.tile(self.controller.model.params.ee_ref, (self.controller.model.params.n_steps + 1 + self.controller.model.params.N, 1)).T
+
+class ReachTargetEXT(AbstractCostSettings):
+    def __init__(self, controller):
+        super().__init__(controller)
+        self.cost_type = 'EXTERNAL'
+        self.set_ocp_cost_type()
+
+        t_glob = self.controller.model.t_glob
+        delta = t_glob - self.controller.ee_params    #self.controller.model.ee_ref
+        track_ee = delta.T @ (self.controller.model.params.Q_weight * np.eye(3)) @ delta 
+        self.controller.ocp.model.cost_expr_ext_cost = track_ee + self.controller.model.u.T @ (self.controller.model.params.R_weight * np.eye(self.controller.model.nu)) @ self.controller.model.u
+        self.controller.ocp.model.cost_expr_ext_cost_e = track_ee
+
+        self.traj = np.tile(self.controller.model.params.ee_ref, (self.controller.model.params.n_steps + 1 + self.controller.model.params.N, 1)).T
```

Comment:
Instead of relying on this parameter, add a method "get_next_reference_traj", which returns the reference trajectory. This allows the user to decide whether the reference trajectory is computed online of before starting the controller (as you are currently assuming).